### DISCUSSION TITLE
Proposal: change const and default definitions

### DISCUSSION THREAD

**Discussion by SorinGFS** [json-schema-org/community#743] [Specification]
the definitions for `const` and `default` actually are:

```json
{
    "const": true,
    "default": true
}
```

I think those definitions are incorrect, because `true` is not a schema. Despite the fact that using `true` would bypass lots of checks.. `true` is not a schema! Alse despite the fact that in various languages `true` also passes a test like `isObject(true)` we here are talking about `json`, and in `json` ... `true` is a `boolean` not an object.

Therefore, I propose to change them to empty object:

```json
{
    "const": {},
    "default": {}
}
```

Their definitions can be found in `meta-data` and `validation`.

This would mean no implications of any kind since empty object `{}` is also a value that evaluates to true.

**Comment by jdesrosiers** [comment]
Boolean values are schemas since draft-06. https://json-schema.org/draft/2020-12/json-schema-core#name-boolean-json-schemas

Personally, I also tend to prefer using `{}` over `true`, but that's a style choice only. `true` is a valid schema.

**Reply by SorinGFS** [reply]
What I meant was from logical point of view, I think when we imagine a schema we think of an object, and in all of the other cases the schema is enclosed in `{ ... }`. Never the less that there are 7 cases in `json-schema` where `"default": {}` is added just because `default` is not what it should be.
Also, all the other keyword definitions may be represented as objects, but when it comes to `const` and `default` this is not possible, because their definition is just a primitive value.

**Reply by jdesrosiers** [reply]
I don't understand what you're trying to say, but I think the problem is that you're misunderstanding how boolean schemas work. The `true` schema means any JSON value is valid. It's equivalent to the empty schema (`{}`). The `false` schema means no JSON value is valid. It's equivalent to `{ "not": {} }`. So, changing `true` to `{}` changes nothing. They behave exactly the same. Which one you use is purely a style choice.

**Reply by notEthan** [reply]
> when we imagine a schema we think of an object

That's the most common sort of schema, but booleans have been valid schemas for a very long time (and have behaved like schemas in a more limited capacity for longer, only as values of `additionalItems` and `additionalProperties`).

> cases in json-schema where "default": {} is added just because default is not what it should be.

The places where `default: {}` is present in the meta-schema aren't related to the schema describing the `default` property being a boolean or an object. They are annotations that basically indicate behavior when the property is omitted is the same as an empty object, corresponding to the text in the specification saying "Omitting this keyword has the same behavior as an empty object."

I'm not quite following what problem changing `true` to `{}` would be solving, but tooling generally needs to be able to handle boolean schemas within any schema, including the meta-schema. If the tool doesn't do that, it's something to be fixed in the tool, not for the meta-schema to change to work around.

**Comment by gregsdennis** [comment]
The two schemas you posted:

```json
{
  "const": true,
  "default": true
}
```
and
```json
{
  "const": {},
  "default": {}
}
```

are not equivalent.  These keywords take JSON values, NOT schemas.

the `true` in `const: true` is merely the JSON literal `true`.  It's not a "true" schema, equivalent to the empty schema `{}`.  Similarly for `default`.

**Reply by SorinGFS** [reply]
@all 
I refer to `const` and `default` keyword definitions not some values allocated to them somewhere, 
here: https://github.com/json-schema-org/json-schema-spec/blob/main/meta/meta-data.json#L16C9-L16C25 and
here: https://github.com/json-schema-org/json-schema-spec/blob/main/meta/validation.json#L20C9-L20C23

**Reply by gregsdennis** [reply]
In that case, I agree with the others. Boolean schemas have been a thing for a _long_ time, and they're well understood.  They're not going away, and it only makes sense to use them.

_You_ don't have to use them, but you do need to understand them because others will continue to use them.

**Reply by SorinGFS** [reply]
@gregsdennis 

> the true in const: true is merely the JSON literal true

in the case of keyword definition I think what you mentioned above in fact is exactly the problem: there is no valid JSON unless is an object, `{}` or `[]`. And a definition at least in my mind, and how is in the rest of 99% keywords, is an object. The content of that object would contain the sum of rules for the keyword it represents, and I think every keyword definition should be representable in its own file, which is what I'm trying to do for a reorganize simulation.

**Reply by notEthan** [reply]
> there is no valid JSON unless is an object, `{}` or `[]`.

This sounds incorrect, unless I'm misunderstanding - `true` is valid JSON, even as the entire contents of a JSON document. If you want to move the definition out and $ref to it, `true` can be the whole contents of a file. The specification defines booleans as valid schemas, regardless of how much or little boolean schemas are used elsewhere in the meta-schema.

**Reply by SorinGFS** [reply]
@notEthan 

> true can be the whole contents of a file

ok, but you can't have a file where you put $id, $schema plus true, like this:

```json
{
   "$id": "https://json-schema.org/draft/next/schema/const",
   "$schema": "https://json-schema.org/draft/next/schema",
   true
}
```

**Reply by notEthan** [reply]
If you want to add $id/$schema you can translate `true` / `false` to their equivalent `{}` / `{"not": {}}` and add to those - but it appears you are already doing that, at least with `true` (and `false` isn't used as a schema in the meta-schema)

**Reply by SorinGFS** [reply]
@notEthan 

As for a keyword definition in json-schema it seem more appropriate `{}` instead of true for those two keywords. That's all, the actual 'true' doesn't bother me.
